compiler/rustc_infer/src/infer/canonical/query_response.rs

```diff
@@ -248,64 +248,71 @@ impl<'tcx> InferCtxt<'tcx> {
     where
         R: Debug + TypeFoldable<TyCtxt<'tcx>>,
     {
-        let InferOk { value: result_args, mut obligations } = self
-            .query_response_instantiation_guess(
-                cause,
-                param_env,
-                original_values,
-                query_response,
-            )?;
+        let result_args =
+            self.query_response_instantiation_guess(cause, original_values, query_response);
+
+        let mut obligations = vec![];
+
+        // Carry all newly resolved opaque types to the caller's scope
+        for &(opaque_type_key, hidden_ty) in &query_response.value.opaque_types {
+            let opaque_type_key = instantiate_value(self.tcx, &result_args, opaque_type_key);
+            let hidden_ty = instantiate_value(self.tcx, &result_args, hidden_ty);
+            debug!(?opaque_type_key, ?hidden_ty, "constrain opaque type");
+            // We can't use equate here, because the hidden type may have been an inference
+            // variable that got constrained to the opaque type itself. In that case we want to ensure
+            // any lifetime differences get recorded in `ouput_query_region_constraints` instead of
+            // being registered in the `InferCtxt`.
+            match hidden_ty.kind() {
+                ty::Alias(ty::Opaque, alias_ty)
+                    if alias_ty.def_id == opaque_type_key.def_id.into() =>
+                {
+                    assert_eq!(alias_ty.args.len(), opaque_type_key.args.len());
+                    for (key_arg, hidden_arg) in
+                        opaque_type_key.args.iter().zip(alias_ty.args.iter())
```

Comment:
This isnt taking into account variance is it?

Reply:
Nope, but that shouldn't matter here in practice, as all the args are either params or inference vars. Well, they should be. Unfortunately in the presence of erroneous code we may not have errored yet, as that only happens at the end of borrowck.